components/fxa-client/src/telemetry.rs

```diff
@@ -4,10 +4,98 @@
 
 pub use crate::http_client::ProfileResponse as Profile;
 use crate::{error::*, FirefoxAccount};
+use serde_derive::*;
+use sync_guid::Guid;
 
 impl FirefoxAccount {
     pub fn get_ecosystem_anon_id(&mut self) -> Result<Option<String>> {
         let profile = self.get_profile(false)?;
         Ok(profile.ecosystem_anon_id)
     }
+
+    /// Gathers and resets telemetry for this account instance.
+    /// This should be considered a short-term solution to telemetry gathering
+    /// and should called whenever consumers expect there might be telemetry,
+    /// and it should submit the telemetry to whatever telemetry system is in
+    /// use (probably glean).
+    ///
+    /// The data is returned as a JSON string, which consumers should parse
+    /// forgivingly (eg, be tolerant of things not existing) to try and avoid
+    /// too many changes as telemetry comes and goes.
+    pub fn gather_telemetry(&mut self) -> Result<String> {
+        let telem = self.telemetry.replace(FxaTelemetry::new());
+        Ok(serde_json::to_string(&telem)?)
+    }
+}
+
+// A somewhat mixed-bag of all telemetry we want to collect. The idea is that
+// the app will "pull" telemetry via a new API whenever it thinks there might
+// be something to record.
+// It's considered a temporary solution until either we can record it directly
+// (eg, via glean) or we come up with something better.
+
+/// The reason a tab/command was received.
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "kebab-case")]
+pub enum ReceivedReason {
+    /// A push notification for the command was received.
+    Push,
+    /// Discovered while handling a push notification for a later message.
+    PushMissed,
+    /// Explicit polling for missed commands.
+    Poll,
+}
+
+#[derive(Debug, Serialize)]
+pub struct SentCommand {
+    pub flow_id: String,
+    pub stream_id: String,
+}
+
+impl Default for SentCommand {
+    fn default() -> Self {
+        Self {
+            flow_id: Guid::random().to_string(),
+            stream_id: Guid::random().to_string(),
+        }
+    }
+}
+
+#[derive(Debug, Serialize)]
+pub struct ReceivedCommand {
+    pub flow_id: String,
+    pub stream_id: String,
+    pub reason: ReceivedReason,
+}
+
+// We have a naive strategy to avoid unbounded memory growth - the intention
+// is that if any platform lets things grow to hit these limits, it's probably
+// never going to consume anything - so it doesn't matter what we discard (ie,
+// there's no good reason to have a smarter circular buffer etc)
```

Comment:
Should we record an event for when the buffer overflows, to see how often this happens? Glean has special `invalid_value` and `invalid_overflow` buckets that it uses when the app records weird stuff (string too long, negative or overflowing counter or enum, unknown discriminant), but I don't know if it's worth doing that here.

Reply:
I don't think it's worth it TBH (at least not now :) Android will grab this immediately after the tab operation, so should never happen. Is this likely on iOS? IOW, do you think we should get an issue opened for this?